pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go

```diff
@@ -0,0 +1,233 @@
+		release, release.GetObjectKind().GroupVersionKind())))
+	// we use partition = 100% to function as "paused" instead of setting pasued field as true
+	// it is manily to keep consistency with partition style (partition is already set as 100% in webhook)
+	patchData.UpdatePaused(false)
+	maxSurge := intstr.FromInt(1) // select the minimum positive number as initial value
+	maxUnavailable := intstr.FromInt(0)
+	patchData.UpdateMaxSurge(&maxSurge)
+	patchData.UpdateMaxUnavailable(&maxUnavailable)
+	patchData.UpdateMinReadySeconds(v1beta1.MaxReadySeconds)
+	klog.InfoS("Initialize: try to update cloneset", "cloneset", klog.KObj(rc.object), "patchData", patchData.String())
+	return rc.client.Patch(context.TODO(), util.GetEmptyObjectWithKey(rc.object), patchData)
+}
+
+func (rc *realController) UpgradeBatch(ctx *batchcontext.BatchContext) error {
+	if err := control.ValidateReadyForBlueGreenRelease(rc.object); err != nil {
+		return errors.NewFatalError(fmt.Errorf("cannot upgrade batch, because cloneset %v doesn't satisfy conditions: %s", klog.KObj(rc.object), err.Error()))
+	}
+	desired, _ := intstr.GetScaledValueFromIntOrPercent(&ctx.DesiredSurge, int(ctx.Replicas), true)
+	current, _ := intstr.GetScaledValueFromIntOrPercent(&ctx.CurrentSurge, int(ctx.Replicas), true)
+	if current >= desired {
+		klog.Infof("No need to upgrade batch for cloneset %v: because current %d >= desired %d", klog.KObj(rc.object), current, desired)
+		return nil
+	} else {
+		klog.Infof("Will update batch for cloneset %v: current %d < desired %d", klog.KObj(rc.object), current, desired)
+	}
+	patchData := patch.NewClonesetPatch()
+	// avoid interference from partition
+	patchData.UpdatePartiton(nil)
+	patchData.UpdateMaxSurge(&ctx.DesiredSurge)
+	return rc.client.Patch(context.TODO(), util.GetEmptyObjectWithKey(rc.object), patchData)
+}
+
+func (rc *realController) Finalize(release *v1beta1.BatchRelease) error {
+	if rc.finalized() {
+		return nil // No need to finalize again
+	}
+
+	if release.Spec.ReleasePlan.BatchPartition != nil {
+		// continuous release (not supported yet)
+		/*
+			patchData := patch.NewClonesetPatch()
+			patchData.DeleteAnnotation(util.BatchReleaseControlAnnotation)
+			return rc.client.Patch(context.TODO(), util.GetEmptyObjectWithKey(rc.object), patchData)
+		*/
+		klog.Warningf("continuous release is not supported yet for bluegreen style release")
+		return nil
+	}
+
+	c := util.GetEmptyObjectWithKey(rc.object)
+	setting := control.GetOriginalSetting(rc.object)
+	patchData := patch.NewClonesetPatch()
+	// why we need a simple MinReadySeconds-based status machine? (ie. the if-else block)
+	// It's possible for Finalize to be called multiple times, if error returned is not nil.
+	// if we do all needed operations in a single code block, like, A->B->C, when C need retry,
+	// both A and B will be executed as well, however, operations like restoreHPA cost a lot(which calls LIST API)
+	if rc.object.Spec.MinReadySeconds == v1beta1.MaxReadySeconds {
```

Comment:
can we compare the spec.minReadySeconds with the one saved in original setting, if they're not match do the patch. Compare minReadySeconds with an arbitrary constant is not a good idea